File: 2018/julia/day10.jl
mutable struct Image
    row::Array{Int}
    col::Array{Int}
    hor_vel::Array{Int}
    ver_vel::Array{Int}
    Image() = new([], [], [], [])
end

function add_point!(self::Image, point::Array{Int})
    push!(self.row, point[1])
    push!(self.col, point[2])
    push!(self.hor_vel, point[3])
    push!(self.ver_vel, point[4])
end

function move!(self::Image, moves=1)
    self.row += self.hor_vel * moves
    self.col += self.ver_vel * moves
end

function compute_area(self::Image)
    abs(reduce(-, extrema(self.row)) * reduce(-, extrema(self.col)))
end

function draw(self::Image)
    row = self.row .- minimum(self.row) .+ 1
    col = self.col .- minimum(self.col) .+ 1
    wid, hei = maximum(row) + 1, maximum(col) + 1
    render = [[' ' for _ in 1:wid] for _ in 1:hei]
    for (i, j) in zip(row, col)
        render[j][i] = '#'
    end
    for row in 1:hei
        println(join(render[row]))
    end
end

function parseLine(str)
    str = replace(str, "position=<" => "")
    str = replace(str, "> velocity=<" => ",")
    str = replace(str, ">" => "")
    [parse(Int, i) for i in split(str, ",")]
end

image = Image()
for line in readlines("data/input-10.txt")
    add_point!(image, parseLine(line))
end

min_area, min_iter = compute_area(image), 0
for it in 1:11000
    move!(image)
    area = compute_area(image)
    if area < min_area
        global min_area, min_iter = area, it
    else
        move!(image, -1)
    end
end

println("Part 1:")
draw(image)

res = min_iter
println("Part 2: $(res)")
@assert(res == 10888)

File: 2018/src/day10.rs
use std::fs;

struct Image {
    row: Vec<isize>,
    col: Vec<isize>,
    hor_vel: Vec<isize>,
    ver_vel: Vec<isize>,
}
impl Image {
    fn new() -> Self {
        Image {
            row: Vec::new(),
            col: Vec::new(),
            hor_vel: Vec::new(),
            ver_vel: Vec::new(),
        }
    }
    fn add_point(&mut self, point: &[isize]) {
        self.row.push(point[0]);
        self.col.push(point[1]);
        self.hor_vel.push(point[2]);
        self.ver_vel.push(point[3]);
    }
    fn move_point(&mut self, moves: isize) {
        for idx in 0..self.row.len() {
            self.row[idx] += self.hor_vel[idx] * moves;
            self.col[idx] += self.ver_vel[idx] * moves;
        }
    }
    fn sizes(&self) -> (usize, usize, isize, isize) {
        let &xmin = self.row.iter().min().unwrap();
        let &ymin = self.col.iter().min().unwrap();
        let wid = self.row.iter().max().unwrap() - xmin;
        let hei = self.col.iter().max().unwrap() - ymin;
        (wid as usize, hei as usize, xmin, ymin)
    }
    fn area(&self) -> usize {
        let (wid, hei, _, _) = self.sizes();
        wid * hei
    }
    fn draw(&self) {
        let (wid, hei, xmin, ymin) = self.sizes();
        let mut render = vec![vec![' '; wid + 1]; hei + 1];
        for (i, j) in self.row.iter().zip(self.col.iter()) {
            render[(j - ymin) as usize][(i - xmin) as usize] = '#';
        }
        for row in render {
            println!("{}", row.iter().collect::<String>());
        }
    }
}

pub fn day10() {
    let input = "data/input-10.txt";
    let file = fs::read_to_string(input)
        .unwrap()
        .replace("position=<", "")
        .replace("> velocity=<", ",")
        .replace('>', "");
    let mut image = Image::new();
    for line in file.lines() {
        let line: Vec<_> = line
            .split(',')
            .map(|z| z.trim().parse::<isize>().unwrap())
            .collect();
        image.add_point(&line);
    }

    let (mut min_area, mut min_iter) = (image.area(), 0);
    for iter in 1..11000 {
        image.move_point(1);
        let area = image.area();
        if area < min_area {
            (min_area, min_iter) = (area, iter);
        } else {
            image.move_point(-1);
            break;
        }
    }

    println!("Part 1:");
    image.draw();

    let res = min_iter;
    println!("Part 2: {res}");
    assert_eq!(res, 10888);
}
